File: BearishEngine/scr/run/Camera.lua
CameraC = class()

function CameraC:new() end

function CameraC:Init()
  self.camera = Camera(mat4():CreateProjection(70 * 3 / 180, 16 / 9, 0.1, 1000),
                      Transform(vec3(0, 0, 0), vec3(1, 1, 1), quat(0, 0, 0, 1)))
end

function CameraC:Update()
  self.camera:SetTransform(self.actor:GetTransform())
end

function CameraC:PreDraw(renderer, camera)
  renderer:SetCamera(self.camera)
end

File: BearishEngine/src/BE/Graphics/Camera.cpp
#include "Camera.h"

using namespace Bearish;
using namespace Graphics;
using namespace Core;
using namespace Math;

Camera* Camera::Identity = new Camera(mat4().CreateIdentity(), Transform());


Camera::Camera(const f32 fov, const f32 aspect, const f32 nearB, const f32 farB, const Core::Transform& transform) : _transform(transform), 
				_projection(mat4().CreateProjection(fov, aspect, nearB, farB)), _near(nearB), _far(farB), _aspect(aspect), _fov(fov) {}

Camera::Camera(const mat4& proj, const Core::Transform& transform) : _transform(transform), _projection(proj) {}


Camera::~Camera() {}

Core::Transform& Camera::GetTransform() {
	return _transform;
}

Math::mat4 Camera::GetProjection() const {
	return _projection;
}

void Camera::SetTransform(const Core::Transform& value) {
	_transform = value;
}

void Camera::SetProjection(const Math::mat4& value) {
	_projection = value;
}

Math::mat4 Camera::GetViewMatrix() const {
	return _projection * _transform.GetCameraTransformation();
}